File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/SqlTemplateMixed.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.Fragment;
import dev.typr.foundations.PgTypes;
import dev.typr.foundations.RowParser;
import dev.typr.foundations.SqlTemplate;
import dev.typr.foundations.Transactor;

import java.sql.SQLException;
import java.util.List;

@SuppressWarnings("unused")
public class SqlTemplateMixed {
    record User(int id, String name, String status) {}

    static RowParser<User> userParser = RowParser.<User>builder()
        .field(PgTypes.int4, User::id)
        .field(PgTypes.text, User::name)
        .field(PgTypes.text, User::status)
        .build(User::new);

    Transactor tx = null; // placeholder

    //start
    // Mix bound and unbound parameters in the same template
    // The status is fixed at "active", but the limit varies per call
    SqlTemplate<Integer, List<User>> activeUsersWithLimit =
        Fragment.of("SELECT id, name, status FROM users WHERE status = ")
            .value(PgTypes.text, "active")
            .append(" ORDER BY name LIMIT ")
            .param(PgTypes.int4)
            .query(userParser.all());

    List<User> topTen() throws SQLException {
        return activeUsersWithLimit.on(10).transact(tx);
    }

    List<User> topFifty() throws SQLException {
        return activeUsersWithLimit.on(50).transact(tx);
    }
    //stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundations/docs/core/SqlTemplateMixed.scala
package dev.typr.foundations.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.data.*


@SuppressWarnings(Array("unused"))
object SqlTemplateMixed:
  case class User(id: Int, name: String, status: String)

  val userParser: RowParser[User] = RowParser.builder[User]()
    .field(PgTypes.int4)(_.id)
    .field(PgTypes.text)(_.name)
    .field(PgTypes.text)(_.status)
    .build(User.apply)

  var tx: Transactor = null // placeholder

  //start
  // Mix bound and unbound parameters in the same template.
  // Status is fixed at "active", limit varies per call.
  val activeUsersWithLimit: SqlTemplate[Int, List[User]] =
    Fragment.of(
      "SELECT id, name, status FROM users WHERE status = "
    ).value(PgTypes.text, "active")
      .append(" ORDER BY name LIMIT ")
      .param(PgTypes.int4)
      .query(userParser.all())

  def topTen(): List[User] =
    activeUsersWithLimit.on(10).transact(tx)

  def topFifty(): List[User] =
    activeUsersWithLimit.on(50).transact(tx)
  //stop
